lib/common/socket.c

```diff
@@ -833,15 +833,12 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
         }
     }
 
-    if (sock->ssl->handshake.client.session_cache) {
-        if (err == NULL || err == h2o_socket_error_ssl_cert_name_mismatch) {
-            SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+    if (sock->ssl->handshake.client.session_cache != NULL) {
+        SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+        if (session != NULL) {
```

Comment:
How about _not_ calling `h2o_cache_set` unless the handshake succeeds?

The intent behind [my previous comment](https://github.com/h2o/h2o/pull/1053#discussion_r76928596) was that the session cache should be updated (either by calling _set_ or _delete_) only after a successful connection establishment.